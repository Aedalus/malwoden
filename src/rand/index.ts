export { AleaRNG } from "./alea";
export { RNG } from "./rng";
