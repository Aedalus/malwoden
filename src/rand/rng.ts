/**
 * An interface for random number generators.
 * Any random number generator from this library will implement the following,
 * and this library can use any custom generator that implements the following.
 */
export interface IRNG {
  next(min?: number, max?: number): number;
  nextInt(min?: number, max?: number): number;
  nextBoolean(): boolean;
  nextItem<T>(array: T[]): T | undefined;
  shuffle<T>(array: T[]): T[];
  reset(): void;
}
