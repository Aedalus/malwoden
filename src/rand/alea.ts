// https://github.com/nquinlan/better-random-numbers-for-javascript-mirror
// Johannes Baag√∏e <baagoe@baagoe.com>, 2010

import { IRNG } from "./rng";

function Mash() {
  var n = 0xefc8249d;

  var mash = function (data: string) {
    data = data.toString();
    for (var i = 0; i < data.length; i++) {
      n += data.charCodeAt(i);
      var h = 0.02519603282416938 * n;
      n = h >>> 0;
      h -= n;
      h *= n;
      n = h >>> 0;
      h -= n;
      n += h * 0x100000000; // 2^32
    }
    return (n >>> 0) * 2.3283064365386963e-10; // 2^-32
  };

  return mash;
}

export class AleaRNG implements IRNG {
  private seed: string;
  private s0 = 0;
  private s1 = 0;
  private s2 = 0;
  private c = 0;

  constructor(seed?: string) {
    // Initialize seed if needed
    this.seed = seed === undefined ? new Date().toString() : seed;
    this.reset();
  }

  reset() {
    let mash = Mash();

    // Initial mashes
    this.s0 = mash(" ");
    this.s1 = mash(" ");
    this.s2 = mash(" ");
    this.c = 1;

    this.s0 -= mash(this.seed);
    this.s1 -= mash(this.seed);
    this.s2 -= mash(this.seed);

    this.sanitize();
  }

  private sanitize() {
    if (this.s0 < 0) this.s0 += 1;
    if (this.s1 < 0) this.s1 += 1;
    if (this.s2 < 0) this.s2 += 1;
  }

  private step(): number {
    // The dark magic that makes the number generator run
    var t = 2091639 * this.s0 + this.c * 2.3283064365386963e-10; // 2^-32
    this.s0 = this.s1;
    this.s1 = this.s2;
    return (this.s2 = t - (this.c = t | 0));
  }

  // Returns between [0,1)
  private nextRand(): number {
    return (
      this.step() + ((this.step() * 0x200000) | 0) * 1.1102230246251565e-16
    ); // 2^-53
  }

  next(min = 0, max = 1): number {
    return this.nextRand() * (max - min) + min;
  }

  nextInt(min = 0, max = 100): number {
    return Math.floor(this.next() * (max - min) + min);
  }

  nextBoolean(): boolean {
    return this.nextRand() > 0.5;
  }

  nextItem<T>(array: T[]): T | undefined {
    if (array.length === 0) return undefined;
    const i = this.nextInt(0, array.length);
    return array[i];
  }

  shuffle<T>(array: T[]): T[] {
    const result: T[] = [];
    const clone = array.slice();
    while (clone.length) {
      const index = this.nextInt(0, clone.length);
      result.push(clone.splice(index, 1)[0]);
    }
    return result;
  }

  clone(): AleaRNG {
    const a = new AleaRNG();
    a.s0 = this.s0;
    a.s1 = this.s1;
    a.s2 = this.s2;
    a.c = this.c;
    return a;
  }
}
