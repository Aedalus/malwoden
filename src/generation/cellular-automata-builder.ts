import { Table } from "../struct";
import { IRNG, AleaRNG } from "../rand";
import { connect, ConnectData } from "./util";
import { Builder } from "./builder";

interface CellularAutomataOptions<T> {
  width: number;
  height: number;
  wallValue: T;
  floorValue: T;
  rng?: IRNG;
}

/** Used to create CellularAutomata Maps. */
export class CellularAutomataBuilder<T> extends Builder<T> {
  private aliveValue: T;
  private deadValue: T;

  private rng: IRNG;

  /**
   * Creates a Cellular Automata Map Generator
   *
   * @param config.width The width of the map.
   * @param config.height The height of the map.
   * @param config.aliveValue The value to use for alive tiles
   * @param config.deadValue The value to use for dead tiles
   */
  constructor(options: CellularAutomataOptions<T>) {
    super(options);

    this.aliveValue = options.wallValue;
    this.deadValue = options.floorValue;
    this.rng = options.rng ?? new AleaRNG();
  }

  /**
   * Randomly sets each cell to either alive or dead.
   *
   * @param isAliveChance The chance for a cell to be set to the 'alive' value.
   */
  randomize(isAliveChance = 0.6) {
    for (let x = 0; x < this.map.width; x++) {
      for (let y = 0; y < this.map.height; y++) {
        const isAlive = this.rng.next() > isAliveChance;
        if (isAlive) {
          this.map.set({ x, y }, this.aliveValue);
        } else {
          this.map.set({ x, y }, this.deadValue);
        }
      }
    }
  }

  private countAliveNeighbors(x: number, y: number): number {
    let count = 0;

    for (let i = -1; i < 2; i++) {
      for (let j = -1; j < 2; j++) {
        let neighbor_x = x + i;
        let neighbor_y = y + j;
        if (i == 0 && j == 0) {
          //this code is supposed to do nothing as it is the focal point we're checking around.
        } else if (
          this.map.isInBounds({ x: neighbor_x, y: neighbor_y }) == false
        ) {
          count++;
        } else if (
          this.map.get({ x: neighbor_x, y: neighbor_y }) == this.aliveValue
        ) {
          count++;
        }
      }
    }
    return count;
  }

  /**
   * Runs a number of simulation steps.
   * Each step generally "smooths" the map.
   *
   * @param stepCount The number of steps to run.
   */
  doSimulationStep(stepCount = 1) {
    for (let step = 0; step < stepCount; step++) {
      const newMap = new Table<T>(this.map.width, this.map.height);
      const oldMap = this.map; //this just renames the table to prevent 'this' all over the place.
      for (let x = 0; x < oldMap.width; x++) {
        for (let y = 0; y < oldMap.height; y++) {
          let nbs = this.countAliveNeighbors(x, y);
          if (oldMap.get({ x, y }) === this.aliveValue) {
            if (nbs < 4) {
              newMap.set({ x, y }, this.deadValue);
            } else {
              newMap.set({ x, y }, this.aliveValue);
            }
          } else {
            if (nbs > 3) {
              newMap.set({ x, y }, this.aliveValue);
            } else {
              newMap.set({ x, y }, this.deadValue);
            }
          }
        }
      }
      this.map = newMap;
    }
  }

  /**
   * Connects areas of the map to ensure they are all connected.
   *
   * For instance, if you're using an alive value of 1 for walls,
   * then this can connect the dead value of 0 to ensure all
   * squares on the map are accessable.
   *
   * @param value The value to connect (default this.deadValue)
   */
  connect(value = this.deadValue): ConnectData {
    return connect(this.map, value);
  }
}
