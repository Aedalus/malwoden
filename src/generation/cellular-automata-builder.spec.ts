import { AleaRNG } from "../rand";
import { CellularAutomataBuilder } from "./cellular-automata-builder";

describe("CellularAutomata", () => {
  it("Will create with an empty table", () => {
    const width = 10;
    const height = 10;

    const a = new CellularAutomataBuilder({
      width,
      height,
      wallValue: 1,
      floorValue: 0,
    });

    for (let x = 0; x < width; x++) {
      for (let y = 0; y < height; y++) {
        expect(a.getMap().get({ x, y })).toBeUndefined();
      }
    }
  });

  it("Will default 1/0 as Alive/Dead values", () => {
    const a = new CellularAutomataBuilder({
      width: 10,
      height: 10,
      wallValue: 1,
      floorValue: 0,
    });
    expect(a["aliveValue"]).toEqual(1);
    expect(a["deadValue"]).toEqual(0);

    const b = new CellularAutomataBuilder({
      width: 10,
      height: 10,
      wallValue: 5,
      floorValue: 4,
    });
    expect(b["aliveValue"]).toEqual(5);
    expect(b["deadValue"]).toEqual(4);
  });

  it("Can randomize with an initial value", () => {
    const width = 10;
    const height = 10;

    const a = new CellularAutomataBuilder({
      width: 10,
      height: 10,
      wallValue: 0,
      floorValue: 1,
      rng: new AleaRNG("foo"),
    });

    a.randomize(0.5);

    let alive = 0;
    let dead = 0;

    const m = a.getMap();
    for (let x = 0; x < width; x++) {
      for (let y = 0; y < height; y++) {
        const val = m.get({ x, y });
        if (val === 1) alive++;
        if (val === 0) dead++;
      }
    }

    expect(alive + dead).toEqual(width * height);
    expect(Math.abs(alive - 50)).toBeLessThanOrEqual(10);
    expect(Math.abs(dead - 50)).toBeLessThanOrEqual(10);
  });

  it("Can run a simulation step", () => {
    const width = 10;
    const height = 10;
    const a = new CellularAutomataBuilder({
      width,
      height,
      wallValue: 1,
      floorValue: 0,
      rng: new AleaRNG("foo"),
    });
    a.randomize();

    a.doSimulationStep();
    expect(a.getMap().items).toMatchSnapshot();
  });

  it("Can connect areas", () => {
    const a = new CellularAutomataBuilder({
      width: 10,
      height: 10,
      wallValue: 1,
      floorValue: 0,
    });
    const m = a.getMap();
    m.fill(1);
    m.set({ x: 3, y: 3 }, 0);
    m.set({ x: 6, y: 3 }, 0);

    const results = a.connect();
    expect(results).toEqual({
      groups: [[{ x: 3, y: 3 }], [{ x: 6, y: 3 }]],
      paths: [
        [
          { x: 6, y: 3, r: 0 },
          { x: 5, y: 3, r: 1 },
          { x: 4, y: 3, r: 2 },
          { x: 3, y: 3, r: 3 },
        ],
      ],
    });
  });
});
