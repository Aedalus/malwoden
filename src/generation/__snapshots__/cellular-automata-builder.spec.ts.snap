// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CellularAutomata Can run a simulation step 1`] = `
Array [
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  0,
  0,
  0,
  1,
  0,
  0,
  0,
  0,
  1,
  1,
  0,
  0,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  0,
  0,
  1,
  1,
  1,
  1,
  0,
  1,
  1,
  1,
  0,
  1,
  1,
  0,
  1,
  1,
  0,
  0,
  0,
  1,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  1,
  1,
  1,
  1,
  0,
  0,
  0,
  0,
  0,
  0,
  1,
  1,
  0,
  0,
  1,
  0,
  0,
  0,
  0,
  0,
  1,
  1,
  1,
  1,
  1,
  0,
  0,
  0,
  0,
  0,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  0,
  1,
]
`;
